cpp/src/traversal/mg/common_utils.cuh

```diff
@@ -102,20 +115,401 @@ struct bfs_frontier_pred_dist {
+                            rmm::device_vector<unsigned> &isolated_bmap)
+{
+  size_t word_count = detail::number_of_words(graph.number_of_vertices);
+  local_isolated_ids.resize(graph.number_of_vertices);
+  global_isolated_ids.resize(graph.number_of_vertices);
+  temp_buffer_len.resize(handle.get_comms().get_size());
+  isolated_bmap.resize(word_count);
+
+  vertex_t local_isolated_count  = populate_isolated_vertices(handle, graph, local_isolated_ids);
+  vertex_t global_isolated_count = collect_vectors(
+    handle, temp_buffer_len, local_isolated_ids, local_isolated_count, global_isolated_ids);
+  add_to_bitmap(handle, isolated_bmap, global_isolated_ids, global_isolated_count);
+}
+
+template <typename T>
+T remove_duplicates(raft::handle_t const &handle, rmm::device_vector<T> &data, T data_len)
+{
+  cudaStream_t stream = handle.get_stream();
+  thrust::sort(rmm::exec_policy(stream)->on(stream), data.begin(), data.begin() + data_len);
+  auto unique_count =
+    thrust::unique(rmm::exec_policy(stream)->on(stream), data.begin(), data.begin() + data_len) -
+    data.begin();
+  return static_cast<T>(unique_count);
+}
+
+// Use the fact that any value in id array can only be in
+// the range [id_begin, id_end) to create a unique set of
+// ids. bmap is expected to be of the length
+// id_end/BitsPWrd<unsigned> and is set to 0 initially
+template <unsigned BLOCK_SIZE, typename T>
+__global__ void remove_duplicates_kernel(
+  unsigned *bmap, T *in_id, T id_begin, T id_end, T count, T *out_id, T *out_count)
+{
+  T tid = blockIdx.x * blockDim.x + threadIdx.x;
+  T id;
+  if (tid < count) {
+    id = in_id[tid];
+  } else {
+    // Invalid vertex id to avoid partial thread block execution
+    id = id_end;
+  }
+
+  int acceptable_vertex = 0;
+  // If id is not in the acceptable range then set it to
+  // an invalid vertex id
+  if ((id >= id_begin) && (id < id_end)) {
+    unsigned active_bit = static_cast<unsigned>(1) << (id % BitsPWrd<unsigned>);
+    unsigned prev_word  = atomicOr(bmap + (id / BitsPWrd<unsigned>), active_bit);
+    // If bit was set by this thread then the id is unique
+    if (!(prev_word & active_bit)) { acceptable_vertex = 1; }
+  }
+
+  __shared__ T block_offset;
+  typedef cub::BlockScan<int, BLOCK_SIZE> BlockScan;
+  __shared__ typename BlockScan::TempStorage temp_storage;
+  int thread_write_offset;
+  int block_acceptable_vertex_count;
+  BlockScan(temp_storage)
+    .ExclusiveSum(acceptable_vertex, thread_write_offset, block_acceptable_vertex_count);
+
+  // If the block is not going to write unique ids then return
+  if (block_acceptable_vertex_count == 0) { return; }
+
+  if (threadIdx.x == 0) {
+    block_offset = cugraph::detail::traversal::atomicAdd(
+      out_count, static_cast<T>(block_acceptable_vertex_count));
+  }
+  __syncthreads();
+
+  if (acceptable_vertex) { out_id[block_offset + thread_write_offset] = id; }
+}
+
+template <unsigned BLOCK_SIZE, typename T>
```

Comment:
Please consider lowercase and more descriptive typenames `<name>_t.` (the comment refers to `typename T`, not the non-type template parameter).